### PR TITLE
REPLACE IMAGE TWEAKS: Fix grid text input, make toggle more logical

### PR DESCRIPTION
**Pull request by ajwl**
## What's changed?

Replace image fields in articles: remove need to hit enter on text input & change the behaviour of the toggle.

Trello: https://trello.com/c/szlvJ9R0/787-replace-image-with-grid-url

With no replacement image present, toggle is hidden:

Screenshot 2019-08-19 at 18 26 29

When replacement image is added, the toggle appears, set to true: 

Screenshot 2019-08-19 at 18 26 52

## Implementation notes

- Knowledge of image replace disappears on save, meaning that the toggle itself will disappear, although it's still possible to remove the image using the little bin icon. 
- Need UX test this flow to confirm whether users understand what to do when the toggle is absent. 

### General
- [ ] 🤖 Relevant tests added
- [ ] ✅ CI checks / tests run locally
- [ ] 🔍 Checked on CODE

### Client
- [ ] 🚫 No obvious console errors on the client (i.e. React dev mode errors)
- [ ] 🎛️ No regressions with existing user interactions (i.e. all existing buttons, inputs etc. work)
- [ ] 📷 Screenshots / GIFs of relevant UI changes included

### CODE REVIEW

**Comment by ajwl** [comment]
@jennygrahamjones fyi!

File: client-v2/src/shared/components/input/InputImage.tsx [hunk 1]
```diff
@@ -289,31 +293,31 @@ class InputImage extends React.Component<ComponentProps, ComponentState> {
 
   private handlePasteImgSrcChange = (e: React.FormEvent<HTMLInputElement>) => {
     e.preventDefault();
-    this.setState({ imageSrc: e.currentTarget.value });
+    e.persist();
```
Comment (ajwl) [inline client-v2/src/shared/components/input/InputImage.tsx]: Check what this does. Do we need it here?

File: client-v2/src/shared/components/input/InputImage.tsx [hunk 2]
```diff
@@ -142,6 +142,9 @@ export interface InputImageContainerProps {
   defaultImageUrl?: string;
   useDefault?: boolean;
   message?: string;
+  replaceImage: boolean;
+  setDisplayImageReplaceToggle: (display: boolean) => void;
```
Comment (jonathonherbert) [inline client-v2/src/shared/components/input/InputImage.tsx]: This couples our image input to the idea of an `imageReplace` field or similar, which future image fields or usages of this component (I'm thinking editions, although there might be other cases) may not need. Could we move this code outside of this component and deal with this logic in the `onChange` handler instead? Happy to talk through or pair on this if needed 👍

Comment (ajwl) [reply client-v2/src/shared/components/input/InputImage.tsx]: Hmm interested by your suggestion. The problem we found is that the imageInput needs to pass certain temporary information (has this image been selected) up to its parent (ArticleParentFormInline - where the toggle is) and this appears to be simplest way to do it. 

Do you mean adding this state into the store in `ImageInput` and then pulling it off in `ArticleParentFormInline`?
I guess we should be cautious about coding around future problems that don't currently exist?

File: client-v2/src/shared/components/input/InputImage.tsx (hunk 1 above)
Comment (jonathonherbert) [reply client-v2/src/shared/components/input/InputImage.tsx]: `persist()` ensures that the event is available outside of its immediate context. It's a bit of an implementation detail from React -- React uses its own synthetic event handlers under the hood, and re-uses event objects liberally, so unless you call `persist()` when passing events elsewhere you can end up in situations where your event has disappeared. That probably sounds like evil magic. It kind of is. See https://reactjs.org/docs/events.html

I can't see that this event is being passed anywhere outside of this function, so you're probably safe to remove. You'll see red ink in the console if there are problems, so I'd remove, test and if all looks good then you're 👍

Comment (jonathonherbert) [reply client-v2/src/shared/components/input/InputImage.tsx]: Just as a pointless NB, other vdom libraries like Preact don't bother with this synthetic event stuff, which I think was largely to smooth out event handling irregularities between browsers. So it doesn't have to be this way! It just _happens_ to be this way for React.

File: client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx [hunk 3]
```diff
@@ -435,23 +439,15 @@ class FormComponent extends React.Component<Props, FormComponentState> {
                       : articleCapiFieldValues.thumbnail
                   }
                   useDefault={!imageCutoutReplace && !imageReplace}
-                  message={imageCutoutReplace ? 'Add cutout' : 'Add image'}
+                  message={imageCutoutReplace ? 'Add cutout' : 'Replace image'}
                   onChange={this.handleImageChange}
+                  setDisplayImageReplaceToggle={
+                    this.setDisplayImageReplaceToggle
+                  }
+                  setImageReplaceToggleValue={this.setImageReplaceToggleValue}
```
Comment (jonathonherbert) [inline client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx]: See below comment re: these new fields -- could we do this in `onChange`, a bit like line 498? (God this form is long)

File: client-v2/src/shared/components/input/InputImage.tsx (hunk 1 above)
Comment (ajwl) [reply client-v2/src/shared/components/input/InputImage.tsx]: cool 👍 thanks for the explanation

**Comment by ajwl** [comment]
Thanks for pointers @jonathonherbert - changes added. 
NB: there is a lingering type error noted above. Should we ignore or try to rewire so it is fixed?

File: client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx [hunk 4]
```diff
@@ -545,14 +550,17 @@ class FormComponent extends React.Component<Props, FormComponentState> {
   };
 
   private handleImageChange: EventWithDataHandler<React.ChangeEvent<any>> = (
-    e,
+    e: ValidationResponse | undefined,
     ...args: [any?, any?, string?]
```
Comment (ajwl) [inline client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx]: What's being passed in here is a `ValidationResponse` object with a src property etc, not an event. What the ConditionalField component expects the signature of the onChange function to be is `EventWithDataHandler<React.ChangeEvent<any>>` : 
```
Type '(e: ValidationResponse | undefined, args_0?: any, args_1?: any, args_2?: string | undefined) => void' is not assignable to type 'EventWithDataHandler<ChangeEvent<any>>'.
  Types of parameters 'e' and 'event' are incompatible.
    Type 'ChangeEvent<any> | undefined' is not assignable to type 'ValidationResponse | undefined'.
      Type 'ChangeEvent<any>' is missing the following properties from type 'ValidationResponse': src, thumb, origin, height, width
```

Comment (jonathonherbert) [reply client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx]: For posterity -- after a chat IRL, this has been resolved by the `unknown` type.

As far as we can tell, redux-form's typings for its `Field` component aren't quite right -- they expect an event in an `onChange` handler, which doesn't match the [current documentation.](https://redux-form.com/7.4.0/docs/api/field.md/#-code-onchange-event-newvalue-previousvalue-name-gt-void-code-optional-) Until they're correct, we patch this by casting the incoming value.

File: client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx [hunk 5]
```diff
@@ -544,15 +549,18 @@ class FormComponent extends React.Component<Props, FormComponentState> {
     return 'primaryImage';
   };
 
+  // onChange: EventWithDataHandler<React.ChangeEvent<any>> & ((_: React.ChangeEvent<any> | undefined) => void)
```
Comment (jonathonherbert) [inline client-v2/src/components/FrontsEdit/ArticleFragmentFormInline.tsx]: We probably don't need this 😁

**Pull request merged**